server/src/main/java/org/elasticsearch/cluster/routing/allocation/IndexMetaDataUpdater.java

```diff
@@ -238,6 +239,11 @@ public static ClusterState removeStaleIdsWithoutRoutings(ClusterState clusterSta
                     }
                     indexMetaDataBuilder.putInSyncAllocationIds(shardNumber, remainingInSyncAllocations);
                 }
+                // Only log the stale shards which have been actually removed.
+                if (oldInSyncAllocations.size() != remainingInSyncAllocations.size()) {
```

Comment:
This check is not necessary. `idsToRemove` already contains the info that's needed (it does not contain non-existing ids, you can even add an assertion for that if you like).
Also I would change the message to "{} marking unavailable shards as stale: {}"

Reply:
I pushed https://github.com/elastic/elasticsearch/pull/28558/commits/b0cd95476487d89efd5d28799e00649a511a6be1